File: Algorithm/Python/1000+/1018. Binary Prefix Divisible By 5.py
# -*- coding: utf-8 -*-
# Created on Sun Apr 07 2019 20:56:32
# Author: WuLC
# EMail: liangchaowu5@gmail.com


# be careful of overflow with too many bits
class Solution(object):
    def prefixesDivBy5(self, A):
        """
        :type A: List[int]
        :rtype: List[bool]
        """
        result = []
        curr = 0
        for num in A:
            curr = (curr * 2 + num) % 5
            if curr == 0:
                result.append(True)
            else:
                result.append(False)
        return result
                

File: Algorithm/C++/1000+/1018. Binary Prefix Divisible By 5.cc
/*
 * Created on Sun Apr 07 2019 20:56:48
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

#include <vector>

using std::vector;

class Solution {
 public:
  vector<bool> prefixesDivBy5(vector<int>& A) {
    vector<bool> result;
    int curr = 0;
    for (auto num : A) {
      curr = (curr * 2 + num) % 5;
      if (curr == 0)
        result.push_back(true);
      else
        result.push_back(false);
    }
    return result;
  }
};